service/storage_proxy.cc

```diff
@@ -3497,8 +3497,40 @@ class abstract_read_executor : public enable_shared_from_this<abstract_read_exec
             });
         });
     }
+    bool maybe_speculatively_fail(digest_read_resolver& resolver, const gms::inet_address& ep, clock_type::time_point timeout) {
+        if (fbu::is_me(ep)) {
+            return false;
+        }
+        if (auto last_seen = _proxy->last_seen(ep)) {
+            auto now = clock_type::now();
+            auto latency_estimate = last_seen->last_sent - last_seen->last_responded;
+            auto time_left = timeout - now;
+            auto last_attempt = now - last_seen->last_sent;
+            // If there was no communication attempt for a long while, the node should be probed in case it came back to life
+            if (last_attempt > time_left) {
+                return false;
```

Comment:
Is the intent that the caller update last_sent, and so the next call will not enter this path?

Maybe the marking and testing should be combined, so this is explicit. a try-style function.